python/tank/util/metrics.py

```diff
@@ -247,44 +270,95 @@ def __init__(self, engine):
         while not self._halt_event.isSet():
 
             # get the next available metric and dispatch it
             try:
-                metrics = MetricsQueueSingleton().get_metrics(
-                    self.DISPATCH_BATCH_SIZE)
-                if metrics:
-                    self._dispatch(metrics)
-            except Exception, e:
+                # For each dispatch cycle, we empty the queue to prevent
+                # metric events from accumulating in the queue.
+                # Because the server has a limit, we dispatch
+                # 'DISPATCH_BATCH_SIZE' items at a time.
+                while True:
+                    metrics = MetricsQueueSingleton().get_metrics(
+                        self.DISPATCH_BATCH_SIZE
+                    )
+                    if metrics:
+                        self._dispatch(metrics)
+                        self._halt_event.wait(self.DISPATCH_SHORT_INTERVAL)
+                    else:
+                        break
+
+            except Exception:
                 pass
             finally:
                 # wait, checking for halt event before more processing
                 self._halt_event.wait(self.DISPATCH_INTERVAL)
 
     def halt(self):
-        """Indiate that the worker thread should halt as soon as possible."""
+        """
+        Ask the worker thread to halt as soon as possible.
+        """
         self._halt_event.set()
 
     def _dispatch(self, metrics):
-        """Dispatch the supplied metric to the sg api registration endpoint.
+        """
+        Dispatch the supplied metric to the sg api registration endpoint and fire
+        the log_metrics hook.
 
-        :param Metric metrics: The Toolkit metric to dispatch.
+        :param metrics: A list of :class:`EventMetric` instances.
+        """
+
+        if self._endpoint_available:
+            self._dispatch_to_endpoint(metrics)
+        # Execute the log_metrics core hook
+        try:
+            self._engine.tank.execute_core_hook_method(
+                constants.TANK_LOG_METRICS_HOOK_NAME,
+                "log_metrics",
+                metrics=[m.data for m in metrics]
+            )
+        except Exception, e:
+            # Catch errors to not kill our thread, log them for debug purpose.
+            self._engine.log_debug("%s hook failed with %s" % (
+                constants.TANK_LOG_METRICS_HOOK_NAME,
+                e,
+            ))
+
+    def _dispatch_to_endpoint(self, metrics):
+        """
+        Dispatch the supplied metric to the sg api registration endpoint. 
 
+        :param metrics: A list of :class:`EventMetric` instances.
         """
 
+        # Filter out metrics we don't want to send to the endpoint.
+        filtered_metrics_data = []
+        for metric in metrics:
+            # Only send internal Toolkit events
+            if metric.is_internal_event:
+                data = metric.data
+                if data["event_name"] not in self.SUPPORTED_EVENTS:
```

Comment:
perhaps we should store the `event_name` as a property so that we don't loose that information? might be useful to evaluate what the "unknown" events are.

Reply:
From the Taxonomy spreadsheet there should be two properties: event name, and event data.  That way we can mine it to see what's getting captured and figure out if we need new official events or some ignore logic.

Reply:
Sorry, totally missed that...